linode/manager
# refactor: [M3-5181] - React Query for Events

## Description 📝

- React Query-ifys our events / events polling system
- This PR's polling system should match the behavior of the existing polling system. The main changes revolve around  the fact that we are now storing and retrieving data from the React Query cache.
- This is take two of https://github.com/linode/manager/pull/9416 because we had to revert it https://github.com/linode/manager/pull/9444
  - The big "bug" with that PR was that `useEventsInfiniteQuery` was mounted many times, which also mounted the cache updating code many times, which caused many renders to happen when event were polled. This PR fixed that by only mounting a polling hook that update the cache once.
  - This PR takes some inspiration from https://github.com/linode/manager/pull/9416, but is different in some ways

## Main Additions

### `useEventsPoller` ⏲️
- A hook mounted at the root of the app, which is responsible for polling the events endpoint for new events
- This hook is responsible for
  - Updating the cache based on the events it polls
  - Showing toast notifications for global toasts (like `volume_delete` for example)
  - Invoking our React Query event handlers (`volumeEventsHandler` for example)

### `useInProgressEvents` 📇
- A hook that returns only active / in-progress events
- Used in `LinodeRow` for example

### `useEventsInfiniteQuery` 📃
- Used to render lists of events
- This is used for `EventsLanding`, `LinodeActivity`, and the Event Menu 

## Other mentions
- Removes `rxjs` dependency 🎉
- Removes events completely from Redux 🗑️

## How to test 🧪
- Verify that all "real-time" feature of Cloud Manager work
  - For example, rebooting a Linode, shutting down a Linode, resizing a Linode, ...
-  Test polling behavior compared to production
   - Verify that this PR does not cause a significant change in the frequency in which we poll the events endpoint
- Test the `http://localhost:3000/events` page
  - Verify the infinite scrolling works as expected
  - Verify the events are in the correct order
  - Verify new events show up in realtime
- Test the Linode Details events tab
  - Verify the infinite scrolling works as expected
  - Verify the events are in the correct order
  - Verify new events show up in realtime
- Test on a large account
- Check for bad performance
- Check the functionality of the Events / Notification Menu in the top navigation
- Check logic and patterns in `packages/manager/src/queries/events.ts` and look for any issues, mistakes, or 🚩
 
## As an Author I have considered 🤔
